keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,992 @@
+    1.  Break ties arbitrarily.
+
+-   An empty `NodeSelector` functions as a default that applies to all nodes.
+    This should be the fall-back and not take precedence if any other range
+    matches. If there are multiple default ranges, ties are broken using the
+    scheme outlined above.
+
+-   `CIDR` and `PerNodeMaskSize` are immutable after creation.
+
+-   The controller will add a finalizer to the ClusterCIDRConfig object when it
+    is created.
+    -   On deletion, the controller checks to see if any Nodes are using
+        `PodCIDRs` from this range -- if so it keeps the finalizer in place and
+        periodically polls Nodes. When all Nodes using this `ClusterCIDRRange`
+        are deleted, the finalizer is removed.
+
+#### Example: Allocations
+
+```go
+[
+  {
+      // For existing clusters this is the same as ClusterCIDR
+      CIDR:            "10.0.0.0/8",
+      // Default for nodes not matching any other rule
+      NodeSelector:    nil,
+      // For existing API this is the same as NodeCIDRMaskSize
+      PerNodeMaskSize: 24,
+  },
+  {
+      CIDR:            "172.16.0.0/14",
+      // Another range, also allocate-able to any node
+      NodeSelector:    nil,
+      PerNodeMaskSize: 24,
+  },
+  {
+      CIDR:            "10.0.0.0/8",
+      NodeSelector:    { key: "np" op: "IN" value:["np1"] },
+      PerNodeMaskSize: 26,
+  },
+  {
+      CIDR:            "192.168.0.0/16",
+      NodeSelector:    { key: "np" op: "IN" value:["np2"] },
+      PerNodeMaskSize: 26,
+  },
+  {
+      CIDR:            "5.2.0.0/16",
+      NodeSelector:    { "np": "np3" },
+      PerNodeMaskSize: 20,
+  },
+  ...
+]
+```
+
+Given the above config, a valid potential configuration might be:
+
+```
+{"np": "np1"} --> "10.0.0.0/26"
+{"np": "np2"} --> "192.16.0.0/26"
+{"np": "np3"} --> "5.2.0.0/20"
+{"np": "np4"} --> "172.16.0.0/24"
+```
+
+### Controller
+
+Implement a new
+[NodeIPAM controller](https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/nodeipam)
+The controller will set up watchers on the `ClusterCIDRConfig` objects and the
+`Node` objects.
+
+This controller relies on being a single writer (just as the current NodeIPAM
+controller does as well). In the case of HA control planes with multiple
+replicas, there will have to be some form of leader election to enforce only 1
+active leader. This KEP proposes re-using the kube-controller-manager leader
+election to pick a active controller.
+
+#### Data Structures
+
+We will use maps to store the allocated ranges and which node is using the
+range. Because the number of nodes is expected to be on the order of thousands,
+more sophisticated data structures are likely not required.
+
+Prior investgations [here](https://github.com/kubernetes/kubernetes/pull/90184)
+suggest that maps storing allocations will perform well under the number of
+nodes we expect.
+
+#### Dual-Stack Support
+
+Before assigning any Pod IPs to nodes, the controller will check all Nodes for
```

Comment:
I think that you want to mean `PodCIDRs` instead of `Pod IPs`

Reply:
yes, good catch. Fixed.